lang/java/avro/src/main/java/org/apache/avro/reflect/ReflectDatumReader.java

```diff
@@ -121,6 +123,18 @@ protected Object readArray(Object old, Schema expected, ResolvingDecoder in)
       @SuppressWarnings("unchecked")
       Collection<Object> c = (Collection<Object>) array;
       return readCollection(c, expectedType, l, in);
+    } else if (array instanceof Map 
+               && ReflectData.isMapWithNonStringKeysSchema(expected)) {
```

Comment:
Sorry, couldn't get you.
For non-string maps, the schema would that be of an array while the object will be of type map.
And the control will come to this only for non-string maps. For string-maps, it will not even come to this.

If you can point me to a case that will fail with this code, I can fix and include the same in test too.